store/postgres/src/notification_listener.rs

```diff
@@ -255,11 +255,27 @@ impl JsonNotification {
 
             // Delete any notifications older than 5 minutes; this serves
             // as a regular cleanup so the table doesn't grow in size indefinitely
-            diesel::sql_query(
-                "DELETE FROM large_notifications
-                 WHERE created_at < current_timestamp - interval '300s'",
-            )
-            .execute(conn)?;
+            //
+            // We make sure that we do not wait for any locks when we do the
+            // cleanup to avoid stalling this transaction. There is not much
+            // danger in this cleanup sometimes leaving rows behind, as long
+            // as they get cleaned up eventually
+
+            // 'delete' locks the table in 'row exclusive' mode; see that we
+            // can get that lock without waiting
+            let lock =
+                diesel::sql_query("lock table large_notifications in row exclusive mode nowait")
+                    .execute(conn);
+            if lock.is_ok() {
+                diesel::sql_query(
+                    "delete from large_notifications
+                     where id in
+                        (select id from large_notifications
+                         where created_at < current_timestamp - interval '300s'
+                         for update skip locked)",
```

Comment:
What does `for update skip locked` do and why do we need both the table lock and this?

Reply:
The code does the locking that `delete` does explicitly: it takes a `row exclusive` lock on the table, and another lock on every row affected by the `delete`. The explicit locking is done in a way that it will never wait for the lock. The `for update` says "lock each selected row for modification" and the `skip locked` says "don't bother with rows that are already locked".

Together, this gives us the effect of "delete what we can without waiting for locks, and leave everything else for somebody else to clean up". All this code is just so that the `large_notifications` table doesn't grow without bounds; it won't really hurt if stuff sticks around for a while longer than 5 minutes.